state-sync/data-streaming-service/src/data_stream.rs

```diff
@@ -483,6 +503,102 @@ impl<T: AptosDataClientInterface + Send + Clone + 'static> DataStream<T> {
+                        .saturating_sub(1) // first_version + num_txns - 1
+                } else {
+                    return Err(aptos_data_client::error::Error::UnexpectedErrorEncountered(
+                        "The first transaction version is missing from the stream response!".into(),
+                    ));
+                }
+            },
+            ResponsePayload::NewTransactionOutputsWithProof((outputs_with_proof, _)) => {
+                if let Some(first_version) = outputs_with_proof.first_transaction_output_version {
+                    let num_outputs = outputs_with_proof.transactions_and_outputs.len();
+                    first_version
+                        .saturating_add(num_outputs as u64)
+                        .saturating_sub(1) // first_version + num_outputs - 1
+                } else {
+                    return Err(aptos_data_client::error::Error::UnexpectedErrorEncountered(
+                        "The first output version is missing from the stream response!".into(),
+                    ));
+                }
+            },
+            _ => {
+                return Ok(()); // The response payload doesn't contain a subscription response
+            },
+        };
+
+        // Get the highest advertised version
+        let highest_advertised_version = global_data_summary
+            .advertised_data
+            .highest_synced_ledger_info()
+            .map(|ledger_info| ledger_info.ledger_info().version())
+            .ok_or(aptos_data_client::error::Error::UnexpectedErrorEncountered(
+                "The highest synced ledger info is missing from the global data summary!".into(),
+            ))?;
+
+        // If the stream is not lagging behind, reset the lag and return
+        if highest_response_version >= highest_advertised_version {
+            self.subscription_stream_lag = None;
+            return Ok(());
+        }
+
+        // Otherwise, the stream is lagging behind the advertised version.
+        // Check if the stream is beyond recovery (i.e., has failed).
+        let current_stream_lag =
+            highest_advertised_version.saturating_sub(highest_response_version);
+        if let Some(mut subscription_stream_lag) = self.subscription_stream_lag.take() {
+            // Check if the stream lag is beyond recovery
+            if subscription_stream_lag
+                .is_beyond_recovery(self.streaming_service_config, current_stream_lag)
+            {
+                return Err(
+                    aptos_data_client::error::Error::SubscriptionStreamIsLagging(format!(
+                        "The subscription stream is beyond recovery! Current lag: {:?}, last lag: {:?},",
+                        current_stream_lag, subscription_stream_lag.version_lag
+                    )),
+                );
+            }
+
+            // The stream is lagging, but it's not yet beyond recovery
+            self.subscription_stream_lag = Some(subscription_stream_lag);
```

Comment:
Would it make sense to have metrics on the lag, both when its "beyond recovery" and even when it's not?

Reply:
Yeah, I think that's a great idea. We already track some of this today (i.e., `lag = advertised - synced`). But, it would be useful to measure this, too. Let me update it 😄